File: src/engine/weapon/projectile.lua
local Actor = require('engine.actor')

---@class Projectile:Actor
---@field speed number
Projectile = setmetatable({}, { __index = Actor })
Projectile.__index = Projectile

---@param spriteEngine SpriteEngine
---@param x number
---@param y number
function Projectile:new(spriteEngine, x, y)
    local obj = Actor:new(spriteEngine, x, y, 0, nil)

    setmetatable(obj, Projectile)
    return obj
end

function Projectile:update(dt)
    self.x = self.x + 20 * dt
    if self.x > love.graphics.getWidth() then
        self.spriteEngine:removeActor(self)
    end
end

function Projectile:draw()
    love.graphics.setColor(255, 0, 0)
    love.graphics.rectangle("fill", self.x, self.y, 1.5, 1.5)
    love.graphics.setColor(255, 255, 255)
end

return Projectile

File: src/engine/weapon/projectile.ts
import { Image } from "love.graphics";
import { config } from "../../conf";
import { Actor } from "../actor";
import { BoxCollider2d } from "../collision/box-collider2d";
import { Engine } from "../engine";
import { SpriteEngine } from "../sprite-engine";
import { AnimatedSprite } from "../sprite/animated-sprite";
import { Rectangle, Vector2 } from "../tools";

let image: Image;
Engine.preload(() => {
  image = love.graphics.newImage("assets/ship-bullet.png");
});

export class Projectile extends Actor {
  constructor(spriteEngine: SpriteEngine, pos: Vector2) {
    const animatedSprite = new AnimatedSprite(image, 16, 16, 0.05);
    const collider = new BoxCollider2d(new Rectangle(5, 5, 10, 6));
    super(spriteEngine, pos, 200, animatedSprite, collider);
  }

  update(dt: number) {
    super.update(dt);
    this.pos.x += this.speed * dt;
    if (this.pos.x > config.screenWidth) {
      this.spriteEngine.removeActor(this);
    }

    // check collisions
    for (const actor of this.spriteEngine.getActors()) {
      if (!actor.collider || (actor as any)["kill"] == undefined) continue;
      if (this.collider?.collides(actor.collider)) {
        (actor as any)["kill"]();
        this.spriteEngine.removeActor(this);
      }
    }
  }
}
